### PR TITLE
Show the date when the script was last run

### PR DESCRIPTION
**Pull request by apoorvaeternity**

### CODE REVIEW

**Comment by henrykironde** [comment]
@apoorvaeternity 

```python
Traceback (most recent call last):
  File "dashboard_script.py", line 3, in <module>
    from datetime import datetime, timezone
ImportError: cannot import name timezone
```

**Comment by apoorvaeternity** [comment]
@henrykironde Are you using Python2? Python3 has timezone in datetime and not Python2.

**Comment by henrykironde** [comment]
Yes I figured that, thanks

**Comment by apoorvaeternity** [comment]
```bash
python manage.py crontab add
  adding cronjob: (a8f524aae7a3acf5108c4df982284a6b) -> ('0 0 * * 0', 'dashboard_script.dashboard_script.run')
```
```bash
python manage.py crontab run a8f524aae7a3acf5108c4df982284a6b
```

**Comment by henrykironde** [comment]
The methods in the `dashboard_script/*` are very long, but I think we shall clean them last.

For now, the error handling needs some revision. 

For example:

```python
    try:
        try:
            dataset_detail = json.load(open('dataset_details.json', 'r'))
        except FileNotFoundError:
            with open("dataset_details.json", 'w') as json_file:
                dataset_detail = dict()
                dataset_detail['dataset_details'] = {}
                json.dump(dataset_detail, json_file)
```
When we fail to find the file during open('dataset_details.json', 'r'), we catch that error.
What happens when we open the file well but we fail to read anything?

Before we try to open it again in` except` , `open("dataset_details.json", 'w') as json_file: `, I think we should close it. Otherwise it may be locked by a process.
```
And I do not think we need to write this file at this point. We are inserting and empty ['dataset_details'] = {} dictionary, we then read it again and try to update it and then we write it back.
To me this seems like double work.
```


What happens when we have the Json file, but it is corrupt?
To me, it seems that it will fail to load and throw an exception which will drive the program to the finally block. In the finally block, we are trying to open the same file after failing.
I may be wrong.

Additionally, I don't clearly understand why we are raising in this part
```python
            try:
                workdir = mkdtemp(dir=file_location)
                download(dataset.name, path=workdir)
                md5 = getmd5(workdir, data_type='dir')
            except Exception:
                raise
```
I feel like that will stop the program from running. 


With Create dirs

```python
    if not os.path.exists(os.path.join(location, 'temp_files')):
        os.makedirs(os.path.join(location, 'temp_files'))
    if not os.path.exists(os.path.join(location, 'old')):
        os.makedirs(os.path.join(location, 'old'))
    if not os.path.exists(os.path.join(location, 'current')):
        os.makedirs(os.path.join(location, 'current'))
    if not os.path.exists(os.path.join(location, 'diffs')):
        os.makedirs(os.path.join(location, 'diffs'))
```
We could later reduce it to three lines

```python
for dir_name in ['temp_files', 'diffs',...]
    if not os.path.exists(os.path.join(location, dir_name)):
        os.makedirs(os.path.join(location, dir_name))
```

We should avoid changing directories if possible.
Since we know the paths to the file location, we can construct paths by joining the file location to the relative paths we want to create.
`file_location = os.path.dirname(os.path.realpath(__file__))`
It is good practice to always normalize the paths, so that running on different platforms is smooth.

@apoorvaeternity overall, looks good but fails to continue whenever some error happens.


Sample run..

```
Installing fao_global_capture_product_unit_data: 100%|██████████████████████████| 7/7 [00:00<00:00, 15575.66rows/s]
Installing fao_global_capture_product_waterarea_groups: 100%|█████████████████| 29/29 [00:00<00:00, 23445.42rows/s]
Installing fao_global_capture_product_dsd_capture_data: 100%|███████████████████| 7/7 [00:00<00:00, 16203.16rows/s]
Installing fao_global_capture_product_species_group: 100%|██████████████| 12723/12723 [00:00<00:00, 13465.54rows/s]
Installing fao_global_capture_product_country_groups: 100%|█████████████████| 275/275 [00:00<00:00, 11019.40rows/s]
Installing fao_global_capture_product_ts_capture_data: 100%|██████████| 932279/932279 [00:28<00:00, 33193.13rows/s]
Downloading 5624799: 2.70kB [00:05, 526B/s]                                                                        
Installing fish_parasite_hosts_host_characteristics: 100%|██████████████| 38008/38008 [00:03<00:00, 12297.36rows/s]
Downloading /.../flensburg-food-web/ECOL_92_174: 160B [00:02, 70.9B/s]                  
Installing flensburg_food_web_links: 100%|████████████████████████████████| 1577/1577 [00:00<00:00, 20151.04rows/s]
Installing flensburg_food_web_nodes: 100%|███████████████████████████████████| 180/180 [00:00<00:00, 8390.66rows/s]
Downloading 5631069: 230B [00:02, 92.1B/s]                                                                         
Installing forest_biomass_china_biomass: 100%|█████████████████████████████| 1626/1626 [00:00<00:00, 5587.14rows/s]
Downloading forestfires.csv: 13.0B [00:00, 71.4B/s]                                                                
Installing forest_fires_portugal_data: 100%|████████████████████████████████| 517/517 [00:00<00:00, 28971.45rows/s]
Downloading REF_SPECIES.csv: 0.00B [00:00, ?B/s]Traceback (most recent call last):
  File "dashboard_script.py", line 78, in <module>
    run()
  File "dashboard_script.py", line 74, in run
    check_dataset(dataset)
  File "dashboard_script.py", line 58, in check_dataset
    json_file_details = json.load(dataset_details_read)
  File "/usr/lib/python3.5/json/__init__.py", line 268, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib/python3.5/json/__init__.py", line 319, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.5/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.5/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
Exception ignored in: <bound method tqdm.__del__ of Downloading REF_SPECIES.csv: 0.00B [00:04, ?B/s]>
Traceback (most recent call last):
  File "...lib/python3.5/site-packages/tqdm/_tqdm.py", line 883, in __del__
    self.close()
  File "..lib/python3.5/site-packages/tqdm/_tqdm.py", line 1088, in close
    self._decr_instances(self)
  File ".../lib/python3.5/site-packages/tqdm/_tqdm.py", line 439, in _decr_instances
    cls._instances.remove(instance)
  File "/usr/lib/python3.5/_weakrefset.py", line 109, in remove
    self.data.remove(ref(item))
KeyError: <weakref at 0x7f9009dbbc28; to 'tqdm' at 0x7f90094aa438>
(my_env36) henr36:~/GitHub/retrieverdash/retrieverdash/dashboard_script$ retriever install sqlite forest-fires-portugal-data
```
Looks like it failed on `forest-inventory-analysis` and I do recommend adding forest inventory to the ignore list and `vertnet`( not the vertnet-*)

**Comment by apoorvaeternity** [comment]
`What happens when we have the Json file, but it is corrupt?` I will handle JSONDecodeError.
```
Additionally, I don't clearly understand why we are raising in this part

            try:
                workdir = mkdtemp(dir=file_location)
                download(dataset.name, path=workdir)
                md5 = getmd5(workdir, data_type='dir')
            except Exception:
                raise

I feel like that will stop the program from running.
```
The error raised will get caught by the outer except block.

File: retrieverdash/dashboard_script/dashboard_script.py [hunk 1]
```diff
@@ -13,24 +14,26 @@
...
+file_location = os.path.normpath(os.path.dirname(os.path.realpath(__file__)))
+
+IGNORE_LIST = ['prism-climate', 'mammal-super-tree', 'forest-inventory-analysis',
+               'vertnet']
 
-IGNORE_LIST = ['prism-climate', 'mammal-super-tree']
 
 def check_dataset(dataset):
-    os.chdir(os.path.join(file_location))
+    os.chdir(file_location)
```
Comment (henrykironde) [inline retrieverdash/dashboard_script/dashboard_script.py]: Why are we changing dir?

File: retrieverdash/dashboard_script/dashboard_script.py [hunk 2]
```diff
@@ -13,24 +14,26 @@
...
     try:
         try:
-            dataset_detail = json.load(open('dataset_details.json', 'r'))
-        except FileNotFoundError:
-            with open("dataset_details.json", 'w') as json_file:
-                dataset_detail = dict()
-                dataset_detail['dataset_details'] = {}
-                json.dump(dataset_detail, json_file)
+            with open("dataset_details.json", 'r') as json_file:
+                dataset_detail = json.load(json_file)
```
Comment (henrykironde) [inline retrieverdash/dashboard_script/dashboard_script.py:35]: This json file gives us the details of all the status for each dataset.
How about we read it in once, before we even enter this function.
Then we pass the data as a dictionary.

When we finish all the precessing of all datasets, the dictionary is written back to the json

Comment (apoorvaeternity) [reply retrieverdash/dashboard_script/dashboard_script.py:35]: The current code is a bit according to the case in which we use multiple processes. I will modify it according to a single thread.

Comment (apoorvaeternity) [reply retrieverdash/dashboard_script/dashboard_script.py:35]: In the current way, we can see the dataset_details.json getting updated after each dataset check.
What would happen if the files are changed during the execution of the script and someone opens the dashboard and tries to access those files?
I think we should write to the dataset_details.json after each dataset is checked.

File: retrieverdash/dashboard_script/dashboard_script.py [hunk 3]
```diff
@@ -13,24 +14,26 @@
...
+        except (FileNotFoundError, JSONDecodeError):
```
Comment (henrykironde) [inline retrieverdash/dashboard_script/dashboard_script.py]: `OSError`
What about IOError, which is now OSError for py3

**Comment by henrykironde** [comment]
Let us keep the commits clean hence let's add this to clean up code commit.

File: retrieverdash/dashboard_script/dashboard_script.py [hunk 4]
```diff
@@ -1,35 +1,38 @@
...
-            dataset_detail = json.load(open('dataset_details.json', 'r'))
-        except FileNotFoundError:
-            with open("dataset_details.json", 'w') as json_file:
-                dataset_detail = dict()
-                json.dump(dataset_detail, json_file)
+            with open("dataset_details.json", 'r') as json_file:
+                dataset_detail = json.load(json_file)
+        except (OSError, JSONDecodeError):
+            dataset_detail = dict()
+            dataset_detail['dataset_details'] = {}
```
Comment (henrykironde) [inline retrieverdash/dashboard_script/dashboard_script.py:38]:
```python
     except Exception as e:
        reason = str(e)	        reason = str(e)
        status = False	        status = False
```
Do we want to type cast this to str()? pyth3 str() is not the same as python2 str()

Comment (apoorvaeternity) [reply retrieverdash/dashboard_script/dashboard_script.py:38]: Yes, I did this it to write to json file. Is there any issue?

**Comment by henrykironde** [comment]
Lets add these to the Ignore list for now.
They are too big
```
biotime
predicts
breed-bird-survey 
predicts
usgs-elevation 
vertnet                       
vertnet-amphibians            
vertnet-birds                 
vertnet-fishes                
vertnet-mammals               
vertnet-reptiles
NPN
```

**Comment by henrykironde** [comment]
I also run out of space and I was wondering where we went wrong. It looks like we have a lot of data. May be we should delete ~/.retriever/raw_data/[script-name] after every script installation.

**Comment by henrykironde** [comment]
Thanks again @apoorvaeternity  for the great work on this.

**Pull request merged**